### PR TITLE
[Inductor UT] Generalize newly introduced inductor UTs for intel GPU (Part 1)

### PR DESCRIPTION
**Pull request by hoshibara**
[Inductor UT] Generalize Newly introduced inductor UTs for intel GPU
reuse `test/inductor/test_autoheuristic.py`
reuse `test/inductor/test_b2b_gemm.py`
reuse `test/inductor/test_custom_lowering.py`
reuse `test/inductor/test_efficient_conv_bn_eval.py`
reuse `test/inductor/test_group_batch_fusion.py`

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @voznesenskym @penguinwu @EikanWang @Guobing-Chen @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @aakhundov

### CODE REVIEW

**Review by etaf — CHANGES_REQUESTED** [review]

File: test/inductor/test_efficient_conv_bn_eval.py [hunk 1]
```diff
@@ -207,17 +207,24 @@ class EfficientConvBNEvalCpuTests(TestCase):
 
     copy_tests(EfficientConvBNEvalTemplate, EfficientConvBNEvalCpuTests, "cpu")
 
-if HAS_CUDA and not TEST_WITH_ASAN:
+if HAS_GPU and not TEST_WITH_ASAN:
+    if GPU_TYPE == "cuda":
 
-    class EfficientConvBNEvalCudaTests(TestCase):
-        device = "cuda"
+        class EfficientConvBNEvalCudaTests(TestCase):
```
[the remark this answers is not in the document]
Comment (hoshibara) [reply test/inductor/test_efficient_conv_bn_eval.py]: Fixed

**Review by etaf — APPROVED** [review]

File: torch/_inductor/fx_passes/b2b_gemm.py [hunk 2]
```diff
@@ -374,7 +374,9 @@ def is_b2b_gemm_good_on(
         B_node.meta["val"],
         C_node.meta["val"],
     )  # torch._subclasses.fake_tensor.FakeTensor
-    if not all([A.is_cuda, B.is_cuda, C.is_cuda]):
+    if not all([A.is_cuda, B.is_cuda, C.is_cuda]) and not all(
+        [A.is_xpu, B.is_xpu, C.is_xpu]
+    ):
```
Comment (EikanWang) [inline torch/_inductor/fx_passes/b2b_gemm.py:379]:
```suggestion
    fake_tensors = (
        A_node.meta["val"],
        B_node.meta["val"],
        C_node.meta["val"],
    )  # torch._subclasses.fake_tensor.FakeTensor
    
    A, B, C = fake_tensors
    
    # Lambda to check if all objects have a specific attribute set to True
    all_have = lambda objects, attr: all(getattr(obj, attr) for obj in objects)
    
    if not all_have(fake_tensors, 'is_cuda') and not all_have(fake_tensors, 'is_xpu'):
```

**Review by EikanWang — CHANGES_REQUESTED** [review]

File: test/inductor/test_b2b_gemm.py [hunk 3]
```diff
@@ -6,10 +6,14 @@
 from torch._inductor.runtime.benchmarking import benchmarker
 from torch._inductor.test_case import run_tests, TestCase
 from torch._inductor.utils import run_and_get_code
-from torch.testing._internal.inductor_utils import HAS_CUDA
+from torch.testing._internal.common_utils import skipIfXpu
+from torch.testing._internal.inductor_utils import GPU_TYPE, HAS_GPU
 
 
+@skipIfXpu(msg="Segmentation fault on CI machine")
```
[the remark this answers is not in the document]
Comment (hoshibara) [reply test/inductor/test_b2b_gemm.py:13]: https://github.com/pytorch/pytorch/issues/138045
File an issue

File: test/inductor/test_custom_lowering.py [hunk 4]
```diff
@@ -59,6 +66,18 @@ def j2pd_cuda(inp, offsets, max_seq_len, pad_value):
...
+                (offsets.shape[0] - 1, max_seq_len, inp.shape[1]),
+                pad_value,
+                device=inp.device,
+                dtype=inp.dtype,
+            )
+            for b in range(offsets.shape[0] - 1):
+                for r in range(offsets[b + 1] - offsets[b]):
+                    res[b][r] = inp[offsets[b] + r]
+            return res
+
```
[the remark this answers is not in the document]
Comment (hoshibara) [reply test/inductor/test_custom_lowering.py]: `j2pd_xpu` and `j2pd_cuda` have been renamed and merge to `j2pd_gpu`.

File: torch/_inductor/fx_passes/b2b_gemm.py (hunk 2 above)
Comment (hoshibara) [reply torch/_inductor/fx_passes/b2b_gemm.py]: Merged

File: test/inductor/test_custom_lowering.py [hunk 5]
```diff
@@ -26,6 +30,9 @@ def setUpClass(cls):
         cls.impl_meta = torch.library.Library(  # noqa: TOR901
             "test_inductor_ops", "IMPL", "Meta"
         )
+        cls.impl_xpu = torch.library.Library(  # noqa: TOR901
+            "test_inductor_ops", "IMPL", "XPU"
+        )
```
[the remark this answers is not in the document]
Comment (hoshibara) [reply test/inductor/test_custom_lowering.py]: Merged

File: test/inductor/test_b2b_gemm.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (EikanWang) [reply test/inductor/test_b2b_gemm.py:13]: So, should we mark it as expectfailure rather than skip it?

[the remark this answers is not in the document]
Comment (hoshibara) [reply test/inductor/test_b2b_gemm.py:13]: expectedFailureXPU cannot block this error message.
Maybe this `Fatal Python error: Segmentation fault` cannot be caught in python level.

**Comment by etaf** [comment]
Please not enable b2b_gemm for now, see https://github.com/pytorch/pytorch/issues/138045

**Review by jansel — APPROVED** [review]

**Review by EikanWang — APPROVED** [review]

**Pull request closed without merging**